app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt

```diff
@@ -100,54 +116,89 @@ object MessageController {
-            getMessagesUidsDelta(mailboxUuid, folderId, previousCursor)
+            getMessagesUidsDelta(mailboxUuid, folder.id, previousCursor)
         } ?: return
 
-        with(messagesUids) {
+        handleMessagesUids(messagesUids, folder, mailboxUuid, previousCursor, threadMode)
+    }
 
-            Log.e("API", "Threads - Added   : ${addedShortUids.count()}")
-            Log.e("API", "Threads - Deleted : ${deletedUids.count()}")
-            Log.e("API", "Threads - Updated : ${updatedMessages.count()}")
+    private suspend fun handleMessagesUids(
+        messagesUids: MessagesUids,
+        folder: Folder,
+        mailboxUuid: String,
+        previousCursor: String?,
+        threadMode: ThreadMode,
+    ) = with(messagesUids) {
 
-            addMessages(addedShortUids, folder, mailboxUuid)
+        Log.i(
+            "API",
+            "Added: ${addedShortUids.count()} | Deleted: ${deletedUids.count()} | Updated: ${updatedMessages.count()} | ${folder.name}",
+        )
 
-            RealmDatabase.mailboxContent().writeBlocking {
+        add(addedShortUids, folder, mailboxUuid, threadMode)
 
-                deleteMessages(deletedUids, realm = this)
-                updateMessages(updatedMessages, mailboxUuid, folderId, realm = this)
+        RealmDatabase.mailboxContent().writeBlocking {
 
-                FolderController.updateFolder(folderId, realm = this) {
-                    it.lastUpdatedAt = Date().toRealmInstant()
-                    it.cursor = cursor
-                }
+            delete(deletedUids, threadMode)
+            update(updatedMessages, folder.id)
+
+            FolderController.getFolder(folder.id, realm = this)?.let {
+                if (previousCursor != null) it.unreadCount = unreadCount
+                it.lastUpdatedAt = Date().toRealmInstant()
+                it.cursor = cursor
             }
         }
-
-        // TODO: Do we still need this with the new API routes?
-        // val isDraftFolder = FolderController.getFolder(folderId, realm = this)?.role == FolderRole.DRAFT
-        // if (isDraftFolder) DraftController.cleanOrphans(threadsResult.threads, realm = this)
     }
 
-    private suspend fun addMessages(shortUids: List<String>, folder: Folder, mailboxUuid: String) {
+    private suspend fun add(
+        shortUids: List<String>,
+        folder: Folder,
+        mailboxUuid: String,
+        threadMode: ThreadMode,
+    ) {
         if (shortUids.isNotEmpty()) {
-            val uids = getUniquesUidsInReverse(folder, shortUids)
-            val pageSize = ApiRepository.PER_PAGE
-            // val pageSize = 200 // TODO: Magic number
+
             var pageStart = 0
+            val pageSize = ApiRepository.PER_PAGE
+            val uids = getUniquesUidsInReverse(folder, shortUids)
+
             while (pageStart < uids.count()) {
+
                 val pageEnd = min(pageStart + pageSize, uids.count())
                 val page = uids.subList(pageStart, pageEnd)
```

Comment:
The `subList` is inclusive with the `fromIndex` and exclusive with the `toIndex`, so we don't have to add a `-1`.

With a `pageSize` of `5`, and without the `-1`, we can see there is no overlapping and no missing `ids`:

<img width="368" alt="image" src="https://user-images.githubusercontent.com/1788629/203924623-70195f21-0972-4507-97a2-351bf6c91b1a.png">

With a `pageSize` of `5`, and with the `-1`, we can see we are missing a fews `ids`:

<img width="316" alt="image" src="https://user-images.githubusercontent.com/1788629/203924770-266e497e-3a27-441a-9239-ad1047705321.png">

Reply:
I already know all that, and precisely the `fromIndex` which is inclusive cause worries here, attention I do not speak about the `pageEnd` but of the `pageStart` but indeed the `pageEnd` there is no need of -1.

For exemple:
>For example for a `pageEnd=50`
> - In the 1st round: `pageSize = 0`
>- In the second round: `pageSize += pageSize`, then `pageStart = 50`.
> And so you will have a `subList(50, ...)` instead of 49.
Isn't that weird?